### PR TITLE
optimize: add WithDialFunc for client

### PR DESCRIPTION
**Pull request by li-jin-gou**
#### What type of PR is this?
optimize
#### Check the PR title.

- [x] This PR title match the format: \<type\>(optional scope): \<description\>
- [x] The description of this PR title is user-oriented and clear enough for others to understand.

#### (Optional) Translate the PR title into Chinese.
Client 添加 WithDialFunc

#### Which issue(s) this PR fixes:
https://github.com/cloudwego/hertz/issues/445#issuecomment-1335468565

### CODE REVIEW

File: pkg/app/client/option.go [hunk 1]
```diff
@@ -147,3 +148,34 @@ func WithWriteTimeout(t time.Duration) config.ClientOption {
 		o.WriteTimeout = t
 	}}
 }
+
+// WithDialFunc is used to set dialer function for DefaultDialer.
+// Deprecated: use WithDialer instead
```
Comment (welkeyever) [inline pkg/app/client/option.go:153]: If decide to provide this api, remove this deprecated tag.

File: pkg/app/client/option.go [hunk 2]
```diff
@@ -147,3 +148,34 @@ func WithWriteTimeout(t time.Duration) config.ClientOption {
...
+func WithDialFunc(f DialFunc) config.ClientOption {
+	return config.ClientOption{F: func(o *config.ClientOptions) {
+		o.Dialer = newCustomDialerWithDialFunc(dialer.DefaultDialer(), f)
```
Comment (welkeyever) [inline pkg/app/client/option.go:156]: What if user also wanna use a customized dialer？

Comment (li-jin-gou) [reply pkg/app/client/option.go:155]: done

File: pkg/app/client/option.go (hunk 1 above)
Comment (li-jin-gou) [reply pkg/app/client/option.go]: done

File: pkg/app/client/option.go (hunk 2 above)
Comment (li-jin-gou) [reply pkg/app/client/option.go]: done

File: pkg/app/client/option.go [hunk 3]
```diff
@@ -147,3 +148,38 @@ func WithWriteTimeout(t time.Duration) config.ClientOption {
...
+	return config.ClientOption{F: func(o *config.ClientOptions) {
+		d := dialer.DefaultDialer()
+		if len(dialers) != 0 {
+			d = dialers[0]
+		}
+		o.Dialer = newCustomDialerWithDialFunc(d, f)
+	}}
+}
+
+type DialFunc func(addr string) (network.Conn, error)
```
Comment (welkeyever) [inline pkg/app/client/option.go:164]: better move it under pkg/network?

Comment (li-jin-gou) [reply pkg/app/client/option.go:164]: thanks

**Review by Duslia — APPROVED** [review]

**Review by FGYFFFF — APPROVED** [review]

**Pull request merged**